### PR TITLE
[Diffusion] Enable Scheduler-Managed Paged KV for HunyuanImage-3.0 DiT

### PR DESCRIPTION
**Pull request by zwhzzz0821**
## Purpose

- Integrate HunyuanImage-3.0 DiT attention with scheduler-managed paged KV, using stable request/CFG row identities to map execution rows to Worker BlockTables.
- Support native FA3 paged attention, mixed causal/full piecewise attention, TP, CFGP, strict Ulysses SP, and grouped step batching.
- Optimize contiguous piecewise segments to avoid unnecessary Q/K/V gathers and output scatter/cat operations. Imported AR KV, Ring SP, AllGather-KV SP, and advanced UAA are out of scope for this PR.

## Test Plan

- Hardware: 4× H200 Hopper GPUs; Python 3.12.13, PyTorch 2.13.0+cu129, vLLM 0.27.0, BF16, and FA3.
- Performance: 1024×1024 images with 8 denoising steps. Run three paired Dense/Paged trials for TP4, TP2+SP2, and TP2+CFGP2, in `D→P`, `P→D`, and `D→P` order. Each case uses one 2-step warmup followed by 30 measured requests. Paged reserves an 8 GiB KV pool.
- Accuracy: Paged scheduler, TP4, FA3, 1024×1024 images, 50 denoising steps, and an 8 GiB KV pool, covering both online and offline pixel-accuracy tests.
- Regression: Run six Hunyuan/paged-KV L1 CPU test files.

```bash
pytest -s -v tests/dfx/perf/scripts/run_diffusion_benchmark.py --test-config-file tests/dfx/perf/tests/test_hunyuan_image_tp4.json
pytest -s -v tests/dfx/perf/scripts/run_diffusion_benchmark.py --test-config-file tests/dfx/perf/tests/test_hunyuan_image_tp2sp2.json
pytest -s -v tests/dfx/perf/scripts/run_diffusion_benchmark.py --test-config-file tests/dfx/perf/tests/test_hunyuan_image_tp2cfgp2.json
pytest -V -s tests/e2e/accuracy/test_hunyuan_image3_pixel_accuracy.py
```

**vLLM Version:** 0.27.0

**vLLM-Omni Commit:** `87325659506444f7d307907f8ea8298a5d7a3b57`

## Test Result

The three performance configurations contain 18 benchmark cases in total. All 540/540 measured requests completed with 0 failures. The table reports the mean ± standard deviation across three trials.

| Topology | Dense E2E | Paged E2E | Latency Change | Dense Throughput | Paged Throughput | Throughput Change |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| TP4 | 2303.667±5.628 ms | 2309.678±8.054 ms | `+0.26%` | 0.434±0.001 req/s | 0.433±0.002 req/s | `-0.26%` |
| TP2+SP2 | 2403.874±10.024 ms | 2381.667±10.571 ms | `-0.92%` | 0.416±0.002 req/s | 0.420±0.002 req/s | `+0.93%` |
| TP2+CFGP2 | 2189.720±9.343 ms | 2211.287±13.014 ms | `+0.98%` | 0.457±0.002 req/s | 0.452±0.003 req/s | `-0.97%` |

Paged grouped batches can continue using FA3 with compact ragged metadata, while the current Dense grouped-batch path requires SDPA to represent the padded 4D mixed-attention mask. Concurrency 4 and 8 completed 12/12 and 16/16 requests respectively, with 0 failures:

| Concurrency | Dense Backend | Paged Backend | Dense QPS | Paged QPS | Throughput Improvement | Dense Mean Latency | Paged Mean Latency | Latency Change |
| ---: | --- | --- | ---: | ---: | ---: | ---: | ---: | ---: |
| 4 | SDPA | FA3 | 0.4192 | 0.4843 | `+15.51%` | 8.8587 s | 7.5914 s | `-14.31%` |
| 8 | SDPA | FA3 | 0.4000 | 0.4650 | `+16.27%` | 18.9374 s | 15.8547 s | `-16.28%` |

### Performance Regression and Improvement Attribution

For single-request runs using the same FA3 backend, the Paged-specific regression comes from page-table indirection inside the FA kernel and writing the current K/V into page slots at every denoising step. After accounting for the Dense KV concatenation eliminated by Paged, the net E2E changes relative to Dense are `+0.26%` latency and `-0.26%` throughput for TP4, and `+0.98%` latency and `-0.97%` throughput for TP2+CFGP2, the worst regression observed. The optimization is that Paged reuses historical KV directly from pages instead of concatenating historical KV with the current K/V at every step; for TP2+SP2, this saving outweighs the page-write and indirection overhead, resulting in `-0.92%` latency and `+0.93%` throughput. For grouped batches, Paged can additionally continue using FA3, while Dense must use SDPA for the padded 4D mixed-attention mask; at concurrency 4/8, this improves throughput by `15.51%`/`16.27%` and reduces latency by `14.31%`/`16.28%`. These percentages are net E2E deltas from paired benchmarks, not per-operator shares measured by separate FA-kernel, KV-write, or concatenation ablations.

Both online and offline pixel-accuracy tests passed (`2 passed, 1 warning in 259.31s`):

| Metric | Worst Observed Value | Threshold | Result |
| --- | ---: | ---: | --- |
| Mean absolute difference | 0.009110293 | ≤ 0.03 | Pass |
| P99 absolute difference | 0.09019607 | ≤ 0.30 | Pass |
| SSIM | 0.981924 | ≥ 0.97 | Pass |
| PSNR | 33.826591 dB | ≥ 30 dB | Pass |

L1 CPU regression: `128 passed, 17 warnings in 6.58s`. Ruff, formatting, SPDX, forbidden-import, `torch.cuda`, and test-marker hooks all passed.

## CC List
@hsliuustc0106 @asukaqaq-s @Bounty-hunter @Gaohan123 @fhfuih @Acerak01-fy

### CODE REVIEW

File: vllm_omni/diffusion/attention/backends/utils/piecewise_attn.py [hunk 1]
```diff
@@ -142,10 +144,11 @@ class PagedPiecewisePlan:
     spans: tuple[tuple[tuple[int, int], ...], ...]
     segments: tuple[PagedPiecewiseSegment, ...]
     num_query_tokens: int
+    segments_cover_query_contiguously: bool
 
 
 PagedPiecewiseRunner = Callable[
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/attention/backends/utils/piecewise_attn.py:150]: [P1] Shared paged-KV seam now diverges from PR #6563 — agree a contract before either merges

#6563 rewrites this alias incompatibly as `[Q, K|None, V|None, meta]` plus a `read_kv_from_cache` mode that skips K/V gathers, vs. `[Q, K, V, meta, output_buffer | None]` here. Overlapping-region rewrites collide in `flash_attn.py::_forward_impl`'s paged branch and `paged_attention_adapter.py` too: output buffer + native-metadata snapshot cloning here, platform write plans (`DiffusionPagedKVWritePlan`) + `paged_query_indices` + output scatter there — plus two parallel Hunyuan integrations (model-forward `_paged_attention_context` vs. runner-driven `pipeline.prepare_paged_attention_rows()`) and different `get_kv_cache_spec` name-canonicalization rules (mandatory prefix vs. getattr fallback in #6563). Whichever lands second needs a semantic port of its Hunyuan delta, not a textual rebase — and the SPDX header conventions currently point in opposite directions (#6563 relaxes the checker while this PR follows main's rewrite-to-Omni policy).

Suggest sequencing these two and extracting one seam contract first: runner protocol, row-builder ownership, and the plan/metadata extension point.

**Comment by Bounty-hunter** [comment]
We need to collect profiling data to understand the root cause of the key performance changes:
1. Whether the attention operator itself shows any performance difference between page input and dense tensor input.
2. Whether the data preparation before attention is more or less expensive under page input. In theory, page input may introduce more read/write operations, while dense input may require an extra concat operation.

File: vllm_omni/diffusion/worker/diffusion_model_runner.py [hunk 2]
```diff
@@ -420,12 +397,26 @@ def get_kv_cache_spec(self) -> dict[str, KVCacheSpec]:
             raise RuntimeError("Model must be loaded before collecting Diffusion KV cache specs")
 
         cache_layers: dict[str, tuple[Attention, KVCacheSpec]] = {}
-        for layer_name, module in self.pipeline.named_modules():
+        cache_layer_paths: dict[str, str] = {}
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/worker/diffusion_model_runner.py:400]: over-defensive? If we want to add this check, using cache_layers directly to detect duplicates should be enough.

File: vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py [hunk 3]
```diff
@@ -1829,18 +1854,42 @@ def _step_group_key(self, state: "StepRequestState") -> tuple[Any, ...]:
...
+            attention_mask = model_kwargs.get("attention_mask")
+            if not isinstance(attention_mask, torch.Tensor) or attention_mask.ndim != 4:
+                raise ValueError(f"Missing Hunyuan attention mask for paged request {state.request_id}.")
+            query_len = int(attention_mask.shape[-2])
+            seq_len = int(attention_mask.shape[-1])
+            full_attn_spans = model_kwargs.get("full_attn_spans")
+            segment_modes = None
+            if isinstance(full_attn_spans, list):
+                segment_modes = tuple(
+                    tuple(segment.mode for segment in build_segments(row_spans, 0, query_len))
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py:1869]: Why does the model/pipeline need to be aware of segment information here? I would expect the model to only pass full_attn_spans through, and let the attention backend build and validate the concrete segments.

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py [hunk 4]
```diff
@@ -1109,32 +1130,47 @@ def forward(
...
-            self.image_kv_cache_lens = None
-            key, value = self._cache_prompt_kv(
-                key,
-                value,
-                seq_len,
-                shard_image_size,
-                kwargs.get("gen_timestep_scatter_index"),
-            )
-            if self.sp_size > 1:
+            if not paged_attention:
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py:1133]: seem to conflct with pr 6563. In hear, it modify forward and add _paged_attention_context. while for 6563, it create new _forward_paged

File: vllm_omni/diffusion/attention/backends/utils/piecewise_attn.py (hunk 1 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/attention/backends/utils/piecewise_attn.py:150]: Following up on the sequencing ask above with the concrete contract worth agreeing on — adopter count isn't hypothetical: 2 models emit `full_attn_spans` today (`hunyuan_image3`, `hidream_o1_image`) and ~11 step-execution pipelines become KV-reuse candidates once continuous batching lands.

Proposal: everything outside `models/<name>/` stays model-blind (scheduler allocation, BlockTables, piecewise planning, platform write-plan hooks already are); per-model cost shrinks to one row-builder file (CFG expansion, prefix source, write-span policy, SP padding rules, capability guards) behind one typed runner-invoked hook — `prepare_paged_attention_rows(states) -> tuple[DiffusionPagedAttentionRow, ...]`, #6563's shape.

Also worth noting: this PR's output-buffer fast path and #6563's K/V-None read-from-cache mode are orthogonal — they merge into one signature `[Q, K|None, V|None, meta, out|None]`; neither has to lose. The real decision is row-builder ownership.

Practical path: land the cleaner lane first and make the second port onto it as a review condition (that port doubles as the generalization proof), then capture an "adopting scheduler-paged KV" doc so HiDream's author copies a doc instead of archaeology across two PRs.

**Comment by zwhzzz0821** [comment]
Closed this PR. The related changes will be consolidated into #6563 going forward.
